ocaml/xapi/xapi_vgpu_type.ml

```diff
@@ -365,27 +365,27 @@ module Vendor_nvidia = struct
   let device_id_of_conf conf = conf.identifier.Identifier.pdev_id
   
   let host_driver_version = try Scanf.sscanf (Unix.readlink !Xapi_globs.nvidia_host_driver_file) "libnvidia-vgpu.so.%s" (fun x -> x)
-    with _ -> info "can not get the file version of the %s, use the default value" !Xapi_globs.nvidia_host_driver_file; Xapi_globs.nvidia_default_host_driver_version
+    with _ -> info "Can not get the file version of %s, use the default value" !Xapi_globs.nvidia_host_driver_file; Xapi_globs.nvidia_default_host_driver_version
 
   let is_host_driver_support_multi_vgpu () =
     List.mem host_driver_version !Xapi_globs.nvidia_multi_vgpu_enabled_driver_versions ||
     (
-      let from_version_on_mark = '+' in
       let host_driver_major_version, host_driver_minor_version = try Scanf.sscanf host_driver_version "%d.%d" (fun x y -> x , y)
-        with _ -> info "get error when parse the host driver version: %s" host_driver_version;
-          Scanf.sscanf default_host_driver_version "%d.%d" (fun x y -> x , y) in (* If cannot parse host driver version, we use the default one,
-                                                                                    which would be taken as not support multiple vGPU *)
+        with _ -> info "Failed to parse Nvidia host driver version: %s" host_driver_version;
+          Scanf.sscanf Xapi_globs.nvidia_default_host_driver_version "%d.%d" (fun x y -> x , y) in (* If cannot parse host driver version, we use the default one,
+                                                                                                      which would be taken as not support multiple vGPU *)
+      let valid_version_pattern = Re.(compile @@ Pcre.re {|([0-9]+)\.([0-9]+)(\+?)|}) in
+      let expected_length = 4 in (* 0. total string, 1. major version, 2. minor version. 3. +*)
       !Xapi_globs.nvidia_multi_vgpu_enabled_driver_versions
-      |> List.filter (fun x -> String.length x > 1)
-      |> List.filter (fun x -> String.get x (String.length x -1) = from_version_on_mark)(* Ending with "+" *)
-      |> List.filter_map ( fun x -> try Scanf.sscanf x "%d.%d" (fun x y -> Some (x , y))
-                           with _ -> None)
-      |> List.exists ( fun (major_version, minor_version) ->
-          match host_driver_major_version - major_version, host_driver_minor_version - minor_version with
-          | _ as x, _ when x > 0 -> true
-          | 0, (_ as y) when y >= 0 -> true
-          | _ -> false
-        )
+      |> List.map (fun x -> Re.Pcre.extract valid_version_pattern x)
```

Comment:
The `extract` can fail with `Not_found` when the string does not match the regular expression - this exception needs to be handled.